### PR TITLE
SONARFLEX-79 Do not use deprecated SonarQube APIs

### PR DESCRIPTION
**Pull request by mpaladin**

### CODE REVIEW

File: sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    // we currently use this hack to remove "import" directives
+    boolean importDirective = false;
+    while (iterator.hasNext()) {
+      Token token = iterator.next();
+      TokenType tokenType = token.getType();
+      if (tokenType == FlexKeyword.IMPORT) {
+        importDirective = true;
+      } else if (importDirective) {
+        // We do nothing as we want to ignore "import" directives
+        if (tokenType == FlexPunctuator.SEMI) {
```
Comment (SonarTech) [inline sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

File: sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  private static String getTokenImage(Token token) {
+    if (token.getType() == GenericTokenType.LITERAL) {
+      return NORMALIZED_CHARACTER_LITERAL;
+    } else if (token.getType() == FlexTokenType.NUMERIC_LITERAL) {
+      return NORMALIZED_NUMERIC_LITERAL;
+    }
+    return token.getValue();
+  }
+
+  private void highlight(NewHighlighting highlighting, Token token, TypeOfText typeOfText) {
```
Comment (SonarTech) [inline sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java]: MAJOR Make "highlight" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      Token token = iterator.next();
+      TokenType tokenType = token.getType();
+      if (tokenType == FlexKeyword.IMPORT) {
+        importDirective = true;
+      } else if (importDirective) {
+        // We do nothing as we want to ignore "import" directives
+        if (tokenType == FlexPunctuator.SEMI) {
+          importDirective = false;
+        }
+      } else if (tokenType != GenericTokenType.EOF) {
```
Comment (SonarTech) [inline sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

File: sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        highlight(highlighting, trivia.getToken(), TypeOfText.COMMENT);
+      }
+    }
+
+    highlighting.save();
+    cpdTokens.save();
+  }
+
+  private static String getTokenImage(Token token) {
+    if (token.getType() == GenericTokenType.LITERAL) {
```
Comment (SonarTech) [inline sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

File: sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+      return NORMALIZED_CHARACTER_LITERAL;
+    } else if (token.getType() == FlexTokenType.NUMERIC_LITERAL) {
```
Comment (SonarTech) [inline sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

File: sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    NewCpdTokens cpdTokens = context.newCpdTokens();
+    cpdTokens.onFile(inputFile);
+
+    Iterator<Token> iterator = lexer.lex(file).iterator();
+    // we currently use this hack to remove "import" directives
+    boolean importDirective = false;
+    while (iterator.hasNext()) {
+      Token token = iterator.next();
+      TokenType tokenType = token.getType();
+      if (tokenType == FlexKeyword.IMPORT) {
```
Comment (SonarTech) [inline sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* MAJOR 6 major

Watch the comments in this conversation to review them.

File: flex-squid/src/main/java/org/sonar/flex/lexer/FlexLexer.java [hunk 7]
```diff
@@ -43,36 +44,49 @@ private FlexLexer() {
...
-        .withChannel(new UnknownCharacterChannel())
+      .withChannel(new UnknownCharacterChannel())
 
-        .build();
+      .build();
+  }
+
+  public static class BomCharacterChannel extends Channel<Lexer> {
+
+    public boolean consume(CodeReader code, Lexer lexer) {
```
Comment (SonarTech) [inline flex-squid/src/main/java/org/sonar/flex/lexer/FlexLexer.java]: MAJOR Add the "@Override" annotation above this method signature [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1161)

File: flex-squid/src/main/java/org/sonar/flex/lexer/FlexLexer.java [hunk 8]
```diff
@@ -43,36 +44,50 @@ private FlexLexer() {
...
-        .withChannel(new UnknownCharacterChannel())
+      .withChannel(new UnknownCharacterChannel())
 
-        .build();
+      .build();
+  }
+
+  private static class BomCharacterChannel extends Channel<Lexer> {
+
+    @Override
```
Comment (SonarTech) [inline flex-squid/src/main/java/org/sonar/flex/lexer/FlexLexer.java]: MAJOR Add the "@Override" annotation above this method signature [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1161)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* MAJOR 5 major

Watch the comments in this conversation to review them.

File: flex-squid/src/main/java/org/sonar/flex/lexer/FlexLexer.java [hunk 9]
```diff
@@ -43,36 +44,50 @@ private FlexLexer() {
...
+      .withChannel(new PunctuatorChannel(FlexPunctuator.values()))
 
-        .withChannel(new UnknownCharacterChannel())
+      .withChannel(new UnknownCharacterChannel())
 
-        .build();
+      .build();
+  }
+
+  private static class BomCharacterChannel extends Channel<Lexer> {
```
Comment (vilchik-elena) [inline flex-squid/src/main/java/org/sonar/flex/lexer/FlexLexer.java]: @mpaladin Why new `BomCharacterChannel`?

Comment (vilchik-elena) [inline flex-squid/src/main/java/org/sonar/flex/lexer/FlexLexer.java]: @mpaladin nothing fails with the old one

Comment (mpaladin) [inline flex-squid/src/main/java/org/sonar/flex/lexer/FlexLexer.java]: @vilchik-elena it's to properly set the column when there is a bom character, there is a failure at runtime with the old one on ruling on `sonar-flex/its/sources/src/as3-commons/as3-commons-asblocks/src/test/actionscript/org/as3commons/asblocks/test/AllTests.as`, with the old one the highlight starts at column 1, it has to start at column 0.

File: sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+import org.sonar.squidbridge.SquidAstVisitor;
+import org.sonar.sslr.parser.LexerlessGrammar;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Set;
+
+public class FlexTokensVisitor extends SquidAstVisitor<LexerlessGrammar> {
```
Comment (vilchik-elena) [inline sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java]: @mpaladin Will `AstAndTokenVisitor` work here?

Comment (mpaladin) [inline sonar-flex-plugin/src/main/java/org/sonar/plugins/flex/FlexTokensVisitor.java]: @vilchik-elena unfortunately not, with the current lexerless grammar there is no type set on tokens, the grammar could probably be modified to set types but that a major change for an upgrade of api.

**Pull request merged**